admin/release.py

```diff
@@ -303,3 +319,178 @@ def publish_docs_main(args, base_path, top_level):
         sys.stderr.write("%s: Can't publish non-tagged version to production."
                          % (base_path.basename(),))
         raise SystemExit(1)
+
+
+class UploadOptions(Options):
+    """
+    Options for uploading packages.
+    """
+    optParameters = [
+        ["target", None, b'clusterhq-yum-repository',
+         "The bucket to upload packages to."],
+        ["build-server", None,
+         b'http://build.clusterhq.com',
+         "The URL of the build-server."],
+    ]
+
+    def parseArgs(self, version):
+        self['version'] = version
+
+
+FLOCKER_PACKAGES = [
+    b'clusterhq-python-flocker',
+    b'clusterhq-flocker-cli',
+    b'clusterhq-flocker-node',
+]
+
+
+def update_repo(rpm_directory, target_bucket, target_key, source_repo,
+                packages):
+    """
+    Update ``target_bucket`` yum repository with ``packages`` from
+    ``source_repo`` repository.
+
+    :param FilePath rpm_directory: Temporary directory to download
+        repository to.
+    :param bytes target_bucket: S3 bucket to upload repository to.
+    :param bytes target_key: Path within S3 bucket to upload repository to.
+    :param bytes source_repo: Repository on the build server to get packages
+        from.
+    :param list packages: List of bytes, each specifying the name of a package
+        to upload to the repository.
+    """
+    # Import these here so that this file can be imported when yum is not
+    # available.
+    from yum import YumBase
+    from pakrat import repo
+
+    rpm_directory.createDirectory()
+    s3 = boto.connect_s3()
+    try:
+        # Does not work if there is a '.' in the name
+        bucket = s3.get_bucket(bucket_name=target_bucket)
+    except boto.exception.S3ResponseError:
+        bucket = s3.create_bucket(bucket_name=target_bucket)
+
+    # Download existing repository
+    for item in bucket.list(prefix=target_key):
```

Comment:
Given the comments below, we perhaps only want to download RPMs here. If we do download metadata, we should keep track of it, so we don't need to upload it again later.